feature/interests/src/main/java/com/google/samples/apps/nowinandroid/feature/interests/InterestsViewModel.kt

```diff
@@ -34,30 +47,98 @@ import javax.inject.Inject
 class InterestsViewModel @Inject constructor(
     val userDataRepository: UserDataRepository,
     getFollowableTopics: GetFollowableTopicsUseCase,
+    userNewsResourceRepository: UserNewsResourceRepository,
+    topicsRepository: TopicsRepository,
+    savedStateHandle: SavedStateHandle,
 ) : ViewModel() {
 
-    val uiState: StateFlow<InterestsUiState> =
-        getFollowableTopics(sortBy = TopicSortField.NAME).map(
-            InterestsUiState::Interests,
-        ).stateIn(
-            scope = viewModelScope,
-            started = SharingStarted.WhileSubscribed(5_000),
-            initialValue = InterestsUiState.Loading,
+    private val topicId: StateFlow<String?> =
+        savedStateHandle.getStateFlow(topicIdArg, null)
+
+    val interestsUiState: StateFlow<InterestsUiState> = combine(
+        getFollowableTopics(sortBy = TopicSortField.NAME),
+        topicId,
+    ) { topics, selectedTopicId ->
+        InterestsUiState.Interests(
+            topics = topics,
+            selectedTopicId = selectedTopicId,
+        )
+    }.stateIn(
+        scope = viewModelScope,
+        started = SharingStarted.WhileSubscribed(5_000),
+        initialValue = InterestsUiState.Loading,
+    )
+
+    val topicUiState: StateFlow<TopicUiState?> = topicId.flatMapLatest { topicId ->
+        topicUiState(
+            topicId,
+            userDataRepository,
+            userNewsResourceRepository,
+            topicsRepository,
         )
+    }.stateIn(
+        scope = viewModelScope,
+        started = SharingStarted.WhileSubscribed(5_000),
+        initialValue = null,
+    )
+
+    fun followTopic(followedTopicId: String, isFollowed: Boolean) {
+        viewModelScope.launch {
+            userDataRepository.toggleFollowedTopicId(followedTopicId, isFollowed)
+        }
+    }
+
+    fun bookmarkNews(newsResourceId: String, isBookmarked: Boolean) {
+        viewModelScope.launch {
+            userDataRepository.updateNewsResourceBookmark(newsResourceId, isBookmarked)
+        }
+    }
 
-    fun followTopic(followedTopicId: String, followed: Boolean) {
+    fun newsViewed(newsResourceId: String) {
         viewModelScope.launch {
-            userDataRepository.toggleFollowedTopicId(followedTopicId, followed)
+            userDataRepository.setNewsResourceViewed(newsResourceId, true)
         }
     }
 }
 
-sealed interface InterestsUiState {
-    object Loading : InterestsUiState
+private fun topicUiState(
+    topicId: String?,
+    userDataRepository: UserDataRepository,
+    userNewsResourceRepository: UserNewsResourceRepository,
+    topicsRepository: TopicsRepository,
+): Flow<TopicUiState?> {
+    if (topicId == null) {
+        return flowOf(null)
```

Comment:
Rx streams don't, but flows do. Otherwise operators like [`filterNotNull`](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/filter-not-null.html) wouldn't make sense